### PR TITLE
Save epoch info in cache

### PR DESCRIPTION
**Pull request by richcarl**
Put the necessary epoch info in the cache for later use.

This PR is supported by Æternity Foundation.

### CODE REVIEW

**Comment by richcarl** [comment]
Rebased on latest master

File: apps/aecore/src/aec_consensus_hc.erl [hunk 1]
```diff
@@ -267,11 +267,20 @@ state_pre_transform_node(Type, Height, PrevNode, Trees) ->
...
     {TxEnv0, _} = aetx_env:tx_env_and_trees_from_hash(aetx_transaction, PrevHash),
     TxEnv = aetx_env:set_height(TxEnv0, Height),
-    {ok, #{epoch := Epoch} = EpochInfo} = aec_chain_hc:epoch_info({TxEnv, Trees}),
+    {ok, #{epoch := Epoch, first := EpochFirst, last := EpochLast} = EpochInfo} =
+        aec_chain_hc:epoch_info({TxEnv, Trees}),
     {ok, Leader} = leader_for_height(Height, {TxEnv, Trees}),
     case Type of
         key ->
-            case Height == maps:get(last, EpochInfo) of
+            case Height =:= EpochFirst of
```
Comment (hanssv) [inline apps/aecore/src/aec_consensus_hc.erl:275]:
```suggestion
            case Height of
```

Comment (hanssv) [reply apps/aecore/src/aec_consensus_hc.erl:275]: and the corresponding changes, would read better?

**Review by hanssv — APPROVED** [review]

File: apps/aecore/src/aec_consensus_hc.erl [hunk 2]
```diff
@@ -290,7 +299,39 @@ state_pre_transform_node(Type, Height, PrevNode, Trees) ->
...
+                     error -> []
+                 end,
+            %% here, the current process must be the leader and is the
+            %% only one who writes to this key, so there is no race
+            aeu_ets_cache:put(?ETS_CACHE_TABLE, child_epoch_start,
+                              {Epoch, [{Epoch, Height, StartTime} | Es]}),
+            ok
+    end.
 
+get_child_epoch_info(Epoch) ->
```
Comment (hanssv) [inline apps/aecore/src/aec_consensus_hc.erl:323]: unused?

Comment (richcarl) [reply apps/aecore/src/aec_consensus_hc.erl:323]: Yes, until a later commit.

File: apps/aecore/src/aec_consensus_hc.erl (hunk 1 above)
Comment (richcarl) [reply apps/aecore/src/aec_consensus_hc.erl]: Turned into ifs instead, maybe clearer.

File: apps/aecore/src/aec_consensus_hc.erl (hunk 2 above)
Comment (hanssv) [reply apps/aecore/src/aec_consensus_hc.erl:322]: Tell that to dialyzer then 😅

File: apps/aecore/src/aec_consensus_hc.erl [hunk 3]
```diff
@@ -267,12 +267,20 @@ state_pre_transform_node(Type, Height, PrevNode, Trees) ->
...
-                true ->
+            if Height =:= EpochFirst ->
+                    %% cache the current epoch start time
+                    EpochStartTime = aetx_env:time_in_msecs(TxEnv),
+                    cache_child_epoch_info(Epoch, Height, EpochStartTime);
+               true ->
+                    ok
+            end,
+            %% note that EpochFirst and EpochLast could be the same, not exclusive
+            if Height =:= EpochLast ->
```
Comment (hanssv) [inline apps/aecore/src/aec_consensus_hc.erl:283]: One if should be enough?

Comment (richcarl) [reply apps/aecore/src/aec_consensus_hc.erl:283]: Both cases need to run independently in case they overlap.

Comment (richcarl) [reply apps/aecore/src/aec_consensus_hc.erl:284]: Probably not, but the code is independent and should not be conflated into a single construct, IMO.

**Review by hanssv — APPROVED** [review]

**Review by kvakvs — APPROVED** [review]

File: apps/aecore/src/aec_consensus_hc.erl [hunk 4]
```diff
@@ -267,12 +268,20 @@ state_pre_transform_node(Type, Height, PrevNode, Trees) ->
...
     TxEnv = aetx_env:set_height(TxEnv0, Height),
-    {ok, #{epoch := Epoch} = EpochInfo} = aec_chain_hc:epoch_info({TxEnv, Trees}),
+    {ok, #{epoch := Epoch, first := EpochFirst, last := EpochLast} = EpochInfo} =
+        aec_chain_hc:epoch_info({TxEnv, Trees}),
     {ok, Leader} = leader_for_height(Height, {TxEnv, Trees}),
     case Type of
         key ->
-            case Height == maps:get(last, EpochInfo) of
-                true ->
+            if Height =:= EpochFirst ->
```
Comment (hanssv) [inline apps/aecore/src/aec_consensus_hc.erl:276]:
```suggestion
            [ cache_child_epoch_info(Epoch, Height, aetx_env:time_in_msecs(TxEnv)) || Height =:= EpochFirst ]
```

Comment (hanssv) [reply apps/aecore/src/aec_consensus_hc.erl:276]: Much clearer then 😛

Comment (richcarl) [reply apps/aecore/src/aec_consensus_hc.erl:276]: Nooooope

**Pull request merged**